examples/user_input.rs

```diff
@@ -47,10 +51,67 @@ impl Default for App {
             input: String::new(),
             input_mode: InputMode::Normal,
             messages: Vec::new(),
+            cursor_in_input: 0,
         }
     }
 }
 
+impl App {
+    fn move_cursor_left(&mut self) {
+        if !self.is_cursor_leftmost() {
+            let cursor_moved_left = self.cursor_in_input - 1;
+            self.cursor_in_input = cursor_moved_left;
+        };
+    }
+
+    fn move_cursor_right(&mut self) {
+        if !self.is_cursor_rightmost() {
+            let cursor_moved_right = self.cursor_in_input + 1;
+            self.cursor_in_input = cursor_moved_right;
+        };
+    }
+
+    fn enter_char_at_cursor(&mut self, new_char: char) {
+        if self.is_cursor_leftmost() {
+            self.input.push(new_char);
+        } else {
+            self.input.insert(self.cursor_in_input, new_char);
+        }
+
+        self.move_cursor_right();
+    }
+
+    fn delete_char_at_cursor(&mut self) {
+        if !self.is_cursor_leftmost() {
+            let current_index = self.cursor_in_input;
+            let from_left_to_current_index = current_index - 1;
+
+            let before_char_to_delete = self.input.chars().take(from_left_to_current_index);
+            let after_char_to_delete = self.input.chars().skip(current_index);
+
+            self.input = before_char_to_delete.chain(after_char_to_delete).collect();
+
+            self.move_cursor_left();
+        }
```

Comment:
```suggestion
        if self.cursor_in_input > 0 {
            self.input.remove(self.cursor_in_input - 1);
        }

        self.move_cursor_left();
```

Reply:
Using remove on Strings in Rust requires to pay attention to the char boundaries. 
If I try you suggestion here, the deletion of chars is quite janky. Sometimes it removes one or two.

Reply:
Oh I didn't think of that nice catch.

Reply:
It's worth documenting this in the code as it's not obvious why this is just by reading the implementation. Also does the following work?
```rust
self.input = self.input.take(position).skip(1).collect();
```

Reply:
```rs
self.input.take(position).skip(1).collect();
```
Does not compile, take is a method of the Itrator trait. String does not implement the Iterator trait.

This would compile:

```rs
self.input = self.input.chars().take(current_index).skip(1).collect();
```

If I understand you correctly then you suggest the following implementation for the function delete_char

```rs
fn delete_char(&mut self) {
    let is_not_cursor_leftmost = self.cursor_position != 0;
    if is_not_cursor_leftmost {
        let current_index = self.cursor_position;
        self.input = self.input.chars().take(current_index).skip(1).collect();
        self.move_cursor_left();
    }
    
}
```

I tested this implementation. It does delete roughly all chars after the selected one. I think that is not intended.

Reply:
Yep you're right - my bad. Just the doc is fine.